### ISSUE TITLE
Reused Input Graph Compiles But Produces Incorrect Output on neuronx-cc 2.10

### ISSUE THREAD

**Issue by xanderdunn** [aws-neuron/aws-neuron-sdk#756]
The graph: 
[rust_hlo_mm_same.pb.zip](https://github.com/aws-neuron/aws-neuron-sdk/files/12744007/rust_hlo_mm_same.pb.zip)

```
debug_ir: HloModule xla_computation_ordered_wrapper, entry_computation_layout={(f32[2,2]{1,0}, f32[2,2]{1,0})->(f32[2,2]{0,1}, f32[2,2]{0,1})}

ENTRY main.11 {
  Arg_0.1 = f32[2,2]{1,0} parameter(0)
  dot.3 = f32[2,2]{0,1} dot(Arg_0.1, Arg_0.1), lhs_contracting_dims={1}, rhs_contracting_dims={0}
  Arg_1.2 = f32[2,2]{1,0} parameter(1)
  transpose.4 = f32[2,2]{0,1} transpose(Arg_0.1), dimensions={1,0}
  dot.5 = f32[2,2]{0,1} dot(Arg_1.2, transpose.4), lhs_contracting_dims={1}, rhs_contracting_dims={0}
  transpose.6 = f32[2,2]{0,1} transpose(Arg_0.1), dimensions={1,0}
  dot.7 = f32[2,2]{0,1} dot(transpose.6, Arg_1.2), lhs_contracting_dims={1}, rhs_contracting_dims={0}
  add.8 = f32[2,2]{0,1} add(dot.5, dot.7)
  ROOT tuple.9 = (f32[2,2]{0,1}, f32[2,2]{0,1}) tuple(dot.3, add.8)
}
```
This graph is pretty simple, if there's anything unusual about it, it's that we are doing a dot product of the input tensor with itself.

This graph successfully compiles, but the outputs are not correct when compared to a Python script that computes the same thing on CPU:
```py
# System
import json

# Third Party
import numpy as np
import torch

def main():
    x_torch = torch.randn(2, 2, dtype=torch.float32, requires_grad=True)
    out_grad_torch = torch.randn(2, 2, dtype=torch.float32)
    out_torch = torch.matmul(x_torch, x_torch)
    assert(out_torch.shape == out_grad_torch.shape)
    out_torch.backward(out_grad_torch)

    res = {
        "x": x_torch.reshape(-1).tolist(),
        "out": out_torch.detach().numpy().reshape(-1).tolist(),
        "x_grad": np.asarray(x_torch.grad).reshape(-1).tolist(),
        "out_grad": out_grad_torch.reshape(-1).tolist(),
        "test1": torch.matmul(out_grad_torch, torch.transpose(x_torch, 0, 1)).reshape(-1).tolist(),
        "test2": torch.matmul(torch.transpose(x_torch, 0, 1), out_grad_torch).reshape(-1).tolist(),
    }

    print(json.dumps(res))


if __name__ == "__main__":
    main()
```
We expect the first output of the XLA graph to be equal to the Python `out` and we expect the second output of the XLA graph to be equal to the Python `x_grad`.

```
$ neuronx-cc --version
NeuronX Compiler version 2.10.0.35+3817a0c8c

Python version 3.8.10
HWM version 2.10.0.5-7b1976adf
NumPy version 1.21.6

Running on AMI ami-01257e71ecb2f431c
Running in region usw2-az4
```

Compiling and running this graph on previous neuronx-cc versions produced the expected output. Compiling and running it on GPU via jax also works for us.

**Comment by aws-taylor** [comment]
Hello @xanderdunn,

 By default, the Neuron compiler will automatically cast FP32 matrix multiplication operations to BF16. I believe you're seeing the consequences of this conversion. I would encourage you to take a look at https://awsdocs-neuron.readthedocs-hosted.com/en/latest/general/appnotes/neuronx-cc/neuronx-cc-training-mixed-precision.html#neuronx-cc-training-mixed-precision for more information on how the Neuron compiler balances performance and accuracy and see if any of the command line options, especially `--auto-cast` and `--auto-cast-type` are able to help you find the sweet spot for performance and accuracy for your application.
 
 You can also find more details about the data types supported on Neuron Cores here: https://awsdocs-neuron.readthedocs-hosted.com/en/latest/general/arch/neuron-features/data-types.html
 
 Please let us know if this help or you have any further questions.

 
 -Taylor

**Comment by xanderdunn** [comment]
Thanks, for our unit tests we compile with this command:
```
neuronx-cc compile /mnt/drive1/tmp/rust_hlo_mm_same.pb --framework=XLA --target=trn1 --model-type=transformer --auto-cast=none --tensorizer-options=--no-run-pg-layout-and-tiling --output /mnt/drive1/tmp/mm_same_8965148480822889077.neff
```

You can see that we explicitly set `--auto-cast=none`.

Let's take an example:

We have inputs:
```
x: [-0.46779287, -0.7243396, -1.9754925, -3.124493]
out_grad: [0.435505, 0.29726982, 0.4401182, 0.037331562]
```

The graph computes `matmul(x, x)` and then the gradient with respect to `out_grad`: `add(matmul(out_grad, x.T), matmul(out_grad, x.T))`

We get outputs:
```
TRN output: [1.6497576, 2.602035, 7.0965333, 11.193383]
jax CPU output: [1.6497576, 2.6020348, 7.0965333, 11.193383]
```
These agree down to 5 decimal places, and the test passes this point.

We get gradients:
```
TRN grad: [-0.09855197, -1.6485941, -0.024820905, -0.85386485]
jax CPU grad: [-1.4922267, -2.0019631, -1.9235246, -1.3180588]
```
These values disagree by up to two orders of magnitude. This is not a minor precision error, the values are completely wrong. When you run this graph with the above two outputs, do you get the same values that I've posted above? Do you get the same values when you run this graph on CPU?

Note that this test was passing on neuronx-cc 2.9. It's also passing when we compile the same graph to CPU or GPU using jax.

**Comment by aws-donkrets** [comment]
Hi [xanderdunn](https://github.com/xanderdunn), there is something else going on besides the --auto-cast interactions. You note that this is a regression compared to the neuronx-cc 2.9 version so we need to take a closer look.

**Comment by xanderdunn** [comment]
Thanks very much @aws-donkrets!

**Comment by aws-donkrets** [comment]
Hi [xanderdunn](https://github.com/xanderdunn), We are testing a potential fix. If successful, this will be resolved in the next SDK release.

**Comment by xanderdunn** [comment]
This reduce is fixed as of `neuronx_cc-2.12.54.0+f631c2365` Thank you! :)

I'll close this issue when it reaches release.

**Comment by xanderdunn** [comment]
This is fixed in 2.16 release
